### ISSUE TITLE
Deep Learning Image: TensorFlow 1.14.0 m33 on Google Cloud produces wrong and non deterministic loss after backpropagation 

### ISSUE THREAD

**Issue by thomasZen** [tensorflow/tensorflow#31797]
**System information**
- Have I written custom code: Yes, the code is attached.
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): `Linux thomas-tf-14 4.9.0-9-amd64 #1 SMP Debian 4.9.168-1+deb9u4 (2019-07-19) x86_64 GNU/Linux`
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: Tested on google cloud
- TensorFlow installed from (source or binary): Binary: `Deep Learning Image: TensorFlow 1.14.0 m33` on Google Cloud
- TensorFlow version: v1.14.0-0-g87989f6 1.14.0
- Python version: 2.7.13 / 3.5.3
- Bazel version (if compiling from source): Not compiled from source
- GCC/Compiler version (if compiling from source): Not compiled from source
- CUDA/cuDNN version: Not used 
- GPU model and memory: None

**Describe the current behavior**
Current behaviour on Google Cloud using `Deep Learning Image: TensorFlow 1.14.0 m33` is non deterministic. During multiple runs the loss after performing backpropagation and updating a variable is different for different runs and does not match the loss of the non optimized standard tensorflow installation. This behaviour exists when using python2 and when using python3.

Created instance with:
`gcloud compute instances create "tf-1-14-cpu" --zone="us-west1-b" --image-family="tf-1-14-cpu" --image-project=deeplearning-platform-release`

Run 1:
```
Loss during step 0: -0.41999998688697815
Loss during step 1: -3.698721931466375e+19
Loss during step 2: -7.38836981337104e+19
```
Run 2:
```
Loss during step 0: -0.41999998688697815
Loss during step 1: -0.41999998688697815
Loss during step 2: -0.41999998688697815
```
Run 3:
```
Loss during step 0: -0.41999998688697815
Loss during step 1: 9.46872814455392e+21
Loss during step 2: 1.8914226722229864e+22
```

**Describe the expected behavior**
On the the same machine using a virtualenv to force the use of non optimized tensorflow as follows:
```
virtualenv -p python3 test
source test/bin/activate
pip3 install tensorflow==1.14.0
```

Run 1:
```
Loss during step 0: -0.41999998688697815
Loss during step 1: -1.4199999570846558
Loss during step 2: -2.4200000762939453
```
Run 2:
```
Loss during step 0: -0.41999998688697815
Loss during step 1: -1.4199999570846558
Loss during step 2: -2.4200000762939453
```
Run 3:
```
Loss during step 0: -0.41999998688697815
Loss during step 1: -1.4199999570846558
Loss during step 2: -2.4200000762939453
```

**Code to reproduce the issue**
```
import tensorflow as tf


if __name__ == "__main__":
    session = tf.Session()
    image = tf.get_variable(name="image", shape=[1, 1, 1, 1], initializer=tf.constant_initializer(0.42))
    session.run(tf.global_variables_initializer())
    kernel = tf.constant(1.0, shape=[1, 1, 1, 1], dtype=tf.float32)

    conv_out = tf.nn.conv2d(image, kernel, strides=[1, 1, 1, 1], padding="SAME")
    max_conv_out = tf.math.reduce_max(conv_out, axis=2)
    loss = -tf.reduce_sum(max_conv_out)
    opt = tf.train.GradientDescentOptimizer(learning_rate=1.0, name="sgd")
    optimizer = opt.minimize(loss, var_list=[image], name="sgd_minimize")

    for step in range(3):
        loss_value, _ = session.run([loss, optimizer])
        print("Loss during step {}: {}".format(step, loss_value))
```

**Other info / logs**
[python-code.txt](https://github.com/tensorflow/tensorflow/files/3520927/python-code.txt)
[tf_env.txt](https://github.com/tensorflow/tensorflow/files/3520903/tf_env.txt)

**Comment by ymodak** [comment]
I tried your code snippet in Google Colab using TF 1.14 and produced same expected results on 3 consecutive runs.
```python
Loss during step 0: -0.41999998688697815
Loss during step 1: -1.4199999570846558
Loss during step 2: -2.4200000762939453
```
Perhaps this is an issue with GCP instance and not with TF 1.14 version specifically.

**Comment by thomasZen** [comment]
@ymodak thanks for running the code in Google Colab. Given these results I assume it used the standard (non mkl optimized) version of TF 1.14. 

Do you have the possibility to test the code with the mkl optimized version of TF 1.14?

**Comment by thomasZen** [comment]
@preethivenkatesh Thanks for checking this. Was it already possible to track down the reason for the non deterministic outputs and in which situations these appear?

**Comment by thomasZen** [comment]
This bug is also present in tensorflow 1.15 with mkl (Deep Learning Image: TensorFlow 1.15.0 m38).

**Comment by preethivenkatesh** [comment]
The issue is resolved and merged into the master branch #35201

**Comment by thomasZen** [comment]
The non deterministic and incorrect loss for the example above resurfaced in tensorflow 1.15. This happens both when installing the python3 library directly and when using docker:
- `pip3 install https://storage.googleapis.com/intel-optimized-tensorflow/intel_tensorflow-1.15.2-cp37-cp37m-manylinux2010_x86_64.whl`
- `docker pull "gcr.io/deeplearning-platform-release/tf-cpu.1-15"`

(CC: https://github.com/tensorflow/tensorflow/pull/35201)

**Comment by preethivenkatesh** [comment]
@thomasZen It looks like we have had perf regression on a few models,  and the commit has been reverted. While we try to resolve this w/o impacting our other critical accelerations, are you okay to use this as a patch for your build?
#35619
